ksql-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -208,36 +230,197 @@ SchemaKTable tableForJoin(
   }
 
 
-  private KsqlTopicSerDe getResultTopicSerde(final PlanNode node) {
-    if (node instanceof StructuredDataSourceNode) {
-      StructuredDataSourceNode structuredDataSourceNode = (StructuredDataSourceNode) node;
-      return structuredDataSourceNode.getStructuredDataSource().getKsqlTopic().getKsqlTopicSerDe();
-    } else if (node instanceof JoinNode) {
-      JoinNode joinNode = (JoinNode) node;
-
-      return getResultTopicSerde(joinNode.getLeft());
-    } else {
-      return getResultTopicSerde(node.getSources().get(0));
+  private void ensureMatchingPartitionCounts(final KafkaTopicClient kafkaTopicClient) {
+    final int leftPartitions = left.getPartitions(kafkaTopicClient);
+    final int rightPartitions = right.getPartitions(kafkaTopicClient);
+    if (leftPartitions != rightPartitions) {
+      throw new KsqlException("Can't join " + leftType.getKqlType() + " with "
```

Comment:
nit: should we use the table/stream names? Or the topic names rather than the type?